### PR TITLE
Alex and Kyle's sections for the report draft

### PR DESCRIPTION
**Pull request by kylrth**
Not sure if Alex is finished, but I've finished my draft for the experiments and results sections.

### CODE REVIEW

File: report/main.tex [hunk 1]
```diff
@@ -52,18 +63,171 @@ \section{Introduction}
...
+A dictionary naturally organizes data into key-value pairs, which would naturally allow for organization and efficient retrieval based on the LLM selecting only relevant keys.
+The LLM would load the information stored at keys it finds relevant to its current task, and organize the information it stores by choosing an appropriate key.
+
+In the context of ART, this translates into the creation of three tools to accomplish our envisioned task. Firstly, we define \texttt{[list-keys]}, a tool that takes no inputs and returns a string containing all the keys in the dictionary.
+Next we define \texttt{[write]}, this tool requires a key-value pair and writes the pair into the dictionary, updating the value stored under the key if it is already present.
+Finally, we define \texttt{[read]}, a tool that takes a key as input and returns the information stored at that key in the dictionary.
+Currently the tools save the dictionary to, and read it from, storage to ensure the long term stability of the data.
+
+We demonstrate our tools by performing a summarization task on journal articles taken from DATASET HERE.
```
Comment (gebelangsn) [inline report/main.tex]: DATASET HERE will be replace by the link to ArXiv summarization?
This is the ArXiv dataset we are loading through the huugingface datasets package:
https://huggingface.co/datasets/ccdv/arxiv-summarization

File: report/main.tex [hunk 2]
```diff
@@ -52,18 +63,171 @@ \section{Introduction}
...
+  \end{subfigure}
+  \caption{Distribution of sentence lengths (number of tokens) in the train and validation sets. For reference, the maximum context sizes for various GPT models are marked as vertical lines. Our method makes it feasible to summarize long documents without access to GPT-4 or GPT-4-32k.}\label{fig:length-hist}
+\end{figure}
+
+To this end, we designed three tools for the model to use: \texttt{[read]}, \texttt{[write]}, and \texttt{[list-keys]}; along with a prompt that explains the task and demonstrates how to use these tools to accomplish the goal of a good summary. The \texttt{[read]} and \texttt{[write]} tools each accept a short text key, and the text associated with the key is stored in a Python dict. This data is specific to each article, and is built by the model as it receives each chunk. The prompt specifically instructs the model to store information under meaningfully-named keys using \texttt{[write]}, and then to \texttt{[read]} all stored information during the last chunk in order to produce the final output.\footnote{See Appendix~\ref{section:prompts} for the text of the prompts used in this project.}
```
Comment (gebelangsn) [inline report/main.tex:111]: are we explaining twice the three tools we designed?

File: report/main.tex [hunk 3]
```diff
@@ -52,18 +63,171 @@ \section{Introduction}
...
+This would allow for the model to dynamically access previously stored information, as well as update its local repository with valuable data as it reads through prompts.
+Given the finite context window of any LLM, ensuring that only relevant data is retrieved would be a crucial requirement of this system.
+With these goals and constraints in mind, we chose a Python dictionary as the underlying data structure for persisting information.
+A dictionary naturally organizes data into key-value pairs, which would naturally allow for organization and efficient retrieval based on the LLM selecting only relevant keys.
+The LLM would load the information stored at keys it finds relevant to its current task, and organize the information it stores by choosing an appropriate key.
+
+In the context of ART, this translates into the creation of three tools to accomplish our envisioned task. Firstly, we define \texttt{[list-keys]}, a tool that takes no inputs and returns a string containing all the keys in the dictionary.
```
Comment (gebelangsn) [inline report/main.tex:81]: are we explaining twice the three tools we designed?

File: report/main.tex [hunk 4]
```diff
@@ -52,18 +63,171 @@ \section{Introduction}
...
-\section{Supplementary Material}
+We observe clear improvement over the base prompt for the two modifications. While the sample size for this experiment was small,\footnote{The cost of this experiment with \texttt{n=10} was US\$16, so we would like more funds in order to validate that these changes improve performance.} we think the difference is significant because they make it possible for more stored information to be \texttt{[read]} when generating the final summary.
 
-Extra stuff to read, if you want.
 
 
+\section{Conclusion}
+
+- context window is limiting to ART framework
+- future work in fine tuning a model instead of using a frozen one.
```
Comment (gebelangsn) [inline report/main.tex]: are these bullet points?

**Review by gebelangsn — APPROVED** [review]
LGTM, but I posted some questions

File: report/main.tex (hunk 3 above)
Comment (kylrth) [reply report/main.tex:81]: Idk, Alex you decide where you prefer the tools to be described

File: report/main.tex [hunk 5]
```diff
@@ -52,18 +63,171 @@ \section{Introduction}
...
+
+
+\section{Methodology}
+%- provide tools that are analogous to long term memory in humans
+
+Our goal was to create a modified ART framework that allows for an LLM to read from, write to, and organize a form of long-term memory, without actually retraining model weights.
+This would allow for the model to dynamically access previously stored information, as well as update its local repository with valuable data as it reads through prompts.
+Given the finite context window of any LLM, ensuring that only relevant data is retrieved would be a crucial requirement of this system.
+With these goals and constraints in mind, we chose a Python dictionary as the underlying data structure for persisting information.
+A dictionary naturally organizes data into key-value pairs, which would naturally allow for organization and efficient retrieval based on the LLM selecting only relevant keys.
```
Comment (kylrth) [inline report/main.tex]: "naturally" is used twice here

File: report/main.tex (hunk 1 above)
Comment (kylrth) [reply report/main.tex]: I discuss the dataset in the Experiments section, so I don't think we need it here.

File: report/main.tex [hunk 6]
```diff
@@ -52,18 +63,171 @@ \section{Introduction}
...
+Next we define \texttt{[write]}, this tool requires a key-value pair and writes the pair into the dictionary, updating the value stored under the key if it is already present.
+Finally, we define \texttt{[read]}, a tool that takes a key as input and returns the information stored at that key in the dictionary.
+Currently the tools save the dictionary to, and read it from, storage to ensure the long term stability of the data.
+
+We demonstrate our tools by performing a summarization task on journal articles taken from DATASET HERE.
+As is standard with ART prompts, demonstrations of the task are prepended to the prompt to provide the LLM an example of how to handle the task~\cite{paranjape2023art}.
+Each article is split into multiple chunks that are fed one at a time into the modified ART framework, with the network choosing how to organize and persist data for later perusal.
+At the end of each article, each entry in the dictionary is then fed back into the LLM and a complete summary is generated. Experiments, results, and their analysis are discussed below.
```
Comment (kylrth) [inline report/main.tex]: I think I expected the methods section to be a little more generic than this. I think I prefer leaving the specifics of the dataset (words like "article" and "summarization") to the Experiments section, and having the Methodology section focus on the theoretical idea of using tools to expand the range of things the model can make use of. Thoughts?

File: report/main.tex (hunk 4 above)
Comment (kylrth) [reply report/main.tex]: Maybe he didn't mean to leave these in

File: report/main.tex [hunk 7]
```diff
@@ -52,18 +63,171 @@ \section{Introduction}
...
 
+\section{Conclusion}
+
+- context window is limiting to ART framework
+- future work in fine tuning a model instead of using a frozen one.
+- lightweight models may be more appropriate
+- Could be used for better conversational AI
+
+In this project we have designed and implemented tools to be used by an LLM to organize and store relevant data for long-term/future use.
+We encountered difficulties in providing enough context for the LLM to properly use the tools, but found some success with simplified run strategies.
```
Comment (kylrth) [inline report/main.tex]:
```suggestion
We encountered difficulties in providing enough context for the LLM to properly use the tools, but found some success by shortening the stored text and the prompt.
```

File: report/main.tex [hunk 8]
```diff
@@ -52,18 +63,171 @@ \section{Introduction}
...
+Future research avenues could involve using newer, bigger, LLMs that have larger context windows allowing for more varied tool use samples to be passed as part of the prompt.
+Otherwise, it may be desirable to fine tune smaller models specifically to use tools and address long standing issues with LLM capabilities.
+To achieve the goal of an AGI system, it will be necessary for it to diversify its problem-solving across many contexts.
```
Comment (kylrth) [inline report/main.tex]:
```suggestion
We'd also like to apply this methodology to make it possible for a frozen LLM to "learn" useful information to an external database, allowing the model to perform continual learning in the same way that it "learned" information as it processed each chunk of the articles in our experiment.
```

File: report/main.tex [hunk 9]
```diff
@@ -52,18 +63,171 @@ \section{Introduction}
...
+Creating systems capable of learning to use tools may be an invaluable step along this path.
```
Comment (kylrth) [inline report/main.tex]:
```suggestion
Creating systems capable of learning to use tools will be an invaluable step in the path to robust models that learn across their deployment lifecycle.
```

Sorry, I don't want to directly attach our project to the idea of AGI, even though that's the name of the class :P

File: report/main.tex (hunk 1 above)
Comment (gebelangsn) [reply report/main.tex]: ohh ok, so line 84 is staying as is?  "We demonstrate our tools by performing a summarization task on journal articles taken from DATASET HERE."
Maybe replacing 'DATASET HERE' by 'ArXiv' sounds more natural?

File: report/main.tex (hunk 6 above)
Comment (AFulleringer) [reply report/main.tex]: To me the theoretical uses of tools would end up in the background section?

Comment (gebelangsn) [reply report/main.tex]: what do you both mean by theoretical uses of tools?
I'll add to the background section a couple of sentences about tools being a way to get around LLM limitations - LLMs are famously bad at math and a tool could help with that...
@kylrth , is there something else Alex should add to Methodology section related to "theoretical idea of using tools " ?

**Pull request merged**